### PR TITLE
Separate authentication from authorization

### PR DESCRIPTION
**Pull request by axw**
## Motivation/summary

Overhaul beater/authorization (now beater/auth) to provide two primary interfaces: Authenticator and Authorizer.

Authenticator is responsible for authenticating a client (typically but not always an agent), returning authentication details (e.g. username for API Key) and an Authorizer for authorizing specific actions.

The auth middleware and interceptor authenticate and store the returned Authorizer in request context, so it can be used by specific method handlers. Authorize accepts an auth.Action and auth.Resource to authorize. When we add anonymous agent auth, the anonymous Authorizer implementation can check the action and use it to decide which properties of the resource to check (i.e. only service name), and which actions to even allow (i.e. disallow source map upload).

HTTP and gRPC status codes are changed to follow the semantics more correctly. Failure to authenticate leads to 401 (HTTP) and Unauthenticated (gRPC). Failure to authorize leads to 403 (HTTP) and PermissionDenied (gRPC). This can only happen when an agent authenticates using an API Key, but the API Key is not authorized for a particular action.

RUM rate limiting now only applies when either secret token or API Key auth is enabled. Otherwise there's no point, backend agents are (currently) never rate limited.

## Checklist

- [x] Update [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/master/CHANGELOG.asciidoc)

## How to test these changes

Mostly non-functional changes. Check auth works as before with the exception of HTTP/gRPC status codes.

## Related issues

https://github.com/elastic/apm-server/issues/5347

### CODE REVIEW

File: beater/auth/authenticator.go [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+
+// NewAuthenticator creates an Authenticator with config, authenticating
+// clients with one of the allowed methods.
+func NewAuthenticator(cfg config.AgentAuth) (*Authenticator, error) {
+	b := Authenticator{secretToken: cfg.SecretToken}
+	if cfg.APIKey.Enabled {
+		// do not use username+password for API Key requests
+		cfg.APIKey.ESConfig.Username = ""
+		cfg.APIKey.ESConfig.Password = ""
+		cfg.APIKey.ESConfig.APIKey = ""
```
Comment (stuartnelson3) [inline beater/auth/authenticator.go:146]: Should apikey be set to `""` if it's being used by the client?

File: beater/auth/authenticator.go [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+			return nil, err
+		}
+
+		cache := newPrivilegesCache(cacheTimeoutMinute, cfg.APIKey.LimitPerMin)
+		b.apikey = newApikeyAuth(client, cache)
+	}
+	return &b, nil
+}
+
+// Authenticate authenticates a client given an authentication mtehod and token,
```
Comment (stuartnelson3) [inline beater/auth/authenticator.go:158]: `/s/mtehod/method/`

File: beater/auth/authenticator.go [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+// returning the authentication details and an Authorizer for authorizing specific
+// actions and resources.
+//
+// Authenticate will return ErrAuthFailed (possibly wrapped) if at least one auth
+// method is configured and the, and no valid credentials have been supplied.
```
Comment (stuartnelson3) [inline beater/auth/authenticator.go:163]: an errant `and the` from a previous version of the comment?

**Review by stuartnelson3 — COMMENTED** [review]
read-through with a few comments

File: beater/auth/apikey.go [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+
+func (c *privilegesCache) get(id string) (*es.HasPrivilegesResponse, bool) {
+	if val, exists := c.cache.Get(id); exists {
+		return val.(*es.HasPrivilegesResponse), true
+	}
+	return nil, false
+}
+
+func (c *privilegesCache) add(id string, privileges *es.HasPrivilegesResponse) {
+	c.cache.SetDefault(id, privileges)
```
[the remark this answers is not in the document]
Comment (axw) [reply beater/auth/apikey.go:201]: I'd like to consolidate the caching techniques we use, and probably use a fixed size LRU here. I think we should probably also rate limit auth attempts. That's for another day though -- this is existing code, I just moved it from `beater/authorization/privilege.go`.

File: beater/auth/authenticator.go (hunk 1 above)
Comment (axw) [reply beater/auth/authenticator.go:146]: This is also existing code, I guess your question comes from the mismatch with the comment above? I'll clarify the comment. The point is to not use any auth defined by apm-server config, and only rely on the API Key provided to the Authenticate method.

File: beater/auth/authenticator.go (hunk 3 above)
Comment (axw) [reply beater/auth/authenticator.go]: astute :)

File: beater/auth/anonymous.go [hunk 5]
```diff
@@ -15,19 +15,16 @@
...
 	"context"
-	"testing"
-
-	"github.com/stretchr/testify/assert"
 )
 
-func TestDenyAuth(t *testing.T) {
-	handler := denyAuth{}
+// AnonymousAuth implements the Authorizer interface.
+type AnonymousAuth struct{}
```
Comment (jalvz) [inline beater/auth/anonymous.go:25]: this is not used right now, correct?

**Review by jalvz — COMMENTED** [review]
An idea for next time to make it smaller might be to separate functional changes from refactor from moved code... It is a bit hard to understand what was copy-pasted and what is new code

File: beater/api/asset/sourcemap/handler_test.go [hunk 6]
```diff
@@ -89,6 +90,20 @@ func TestAssetHandler(t *testing.T) {
...
 			code: http.StatusAccepted,
 		},
+		"unauthorized": {
+			authorizer: func(context.Context, auth.Action, auth.Resource) error {
+				return auth.ErrUnauthorized
+			},
+			code: http.StatusForbidden,
+			body: beatertest.ResultErrWrap("unauthorized"),
+		},
+		"auth_unavailable": {
```
[the remark this answers is not in the document]
Comment (axw) [reply beater/api/asset/sourcemap/handler_test.go:100]: The point of that test is to point out that errors returned by Authenticate other than ErrAuthFailed do not result in 401. i.e. it's not the client's fault, there's something wrong with the server/backend.

File: beater/auth/anonymous.go (hunk 5 above)
Comment (axw) [reply beater/auth/anonymous.go:25]: It's used in the auth middleware when auth isn't required, and in the Jaeger method authenticator for agent config.

**Comment by axw** [comment]
> An idea for next time to make it smaller might be to separate functional changes from refactor from moved code... It is a bit hard to understand what was copy-pasted and what is new code

Yep fair comment. I had figured there wasn't that much code that was just moved to warrant that, but evidently there was enough that it could have helped to separate. Thanks!

**Review by simitt — APPROVED** [review]
This separation is another great improvement.

**Comment by stuartnelson3** [comment]
Confirmed a 401 for HTTP and code = Unauthenticated for gRPC with BC2, request was successful when the correct auth was supplied

**Pull request merged**